src/Compilers/CSharp/Test/Emit2/Semantics/CollectionLiteralTests.cs

```diff
@@ -5644,6 +5657,2663 @@ private static void VerifyTypes(SemanticModel model, ExpressionSyntax expr, stri
+        {
+            string sourceA = """
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), "Create")]
+                public struct MyCollection<T> : IEnumerable<T>
+                {
+                    public IEnumerator<T> GetEnumerator() => default;
+                    IEnumerator IEnumerable.GetEnumerator() => default;
+                }
+                public class MyCollectionBuilder
+                {
+                    public static void Create(int[] items) { }
+                }
+                """;
+            var comp = CreateCompilation(new[] { sourceA, CollectionBuilderAttributeDefinition }, targetFramework: TargetFramework.Mscorlib40);
+            comp.VerifyEmitDiagnostics();
+            var refA = comp.ToMetadataReference();
+
+            string sourceB = """
+                #pragma warning disable 219
+                class Program
+                {
+                    static void Main()
+                    {
+                        F([]);
+                        F(new());
+                    }
+                    static void F(MyCollection<int> c)
+                    {
+                    }
+                }
+                """;
+            comp = CreateCompilation(sourceB, references: new[] { refA }, targetFramework: TargetFramework.Mscorlib45);
+            comp.VerifyEmitDiagnostics(
+                // (6,11): error CS9187: Could not find an accessible 'Create' method with the expected signature: a static method with a single parameter of type 'ReadOnlySpan<T>' and return type 'MyCollection<T>'.
+                //         F([]);
+                Diagnostic(ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound, "[]").WithArguments("Create", "T", "MyCollection<T>").WithLocation(6, 11));
+
+            var collectionType = (NamedTypeSymbol)comp.GetMember<MethodSymbol>("Program.F").Parameters[0].Type;
+            Assert.Equal("MyCollection<System.Int32>", collectionType.ToTestDisplayString());
+            TypeSymbol builderType;
+            string methodName;
+            Assert.True(collectionType.HasCollectionBuilderAttribute(out builderType, out methodName));
+            Assert.Equal("MyCollectionBuilder", builderType.ToTestDisplayString());
+            Assert.Equal("Create", methodName);
+
+            var retargetingType = (RetargetingNamedTypeSymbol)collectionType.OriginalDefinition;
+            Assert.Equal("MyCollection<T>", retargetingType.ToTestDisplayString());
+            Assert.True(retargetingType.HasCollectionBuilderAttribute(out builderType, out methodName));
+            Assert.IsType<RetargetingNamedTypeSymbol>(builderType);
+            Assert.Equal("MyCollectionBuilder", builderType.ToTestDisplayString());
+            Assert.Equal("Create", methodName);
+        }
+
```

Comment:
Consider testing applying the attribute to a base collection type then deriving from it. Maybe including cases where the attribute is improperly marked Inherited=true.

Consider testing a case where the builder type is some derived type, and the Create method is declared on the base type of the builder type.

Consider testing ref struct collection types with and without 'scoped' ROS Create parameters. Can mark with PROTOTYPE assuming the optimized codegen in the scoped case is not urgent.

Reply:
Added suggested tests, thanks.